File: src/#654/solution.js
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val
  this.left = left === undefined ? null : left
  this.right = right === undefined ? null : right
}

var constructMaximumBinaryTree = function (nums) {
  function buildTree(nums = [], start = 0, end = 0) {
    if (nums.length == 0) {
      return null
    }
    let maxIndex = 0
    nums.forEach((v, i) => {
      if (v > nums[maxIndex]) {
        maxIndex = i
      }
    })
    let root = new TreeNode(nums[maxIndex], null, null)
    root.left = buildTree(nums.slice(0, maxIndex), start, maxIndex - 1)
    root.right = buildTree(nums.slice(maxIndex + 1, nums.length - 1), maxIndex + 1, end)
    return root
  }

  return buildTree(nums, 0, nums.length - 1)
}

File: src/#654/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func constructMaximumBinaryTree(nums []int) *TreeNode {

	return buildTree(nums, 0, len(nums)-1)
}

func buildTree(nums []int, start, end int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	maxIndex := 0
	for i, v := range nums {
		if v > nums[maxIndex] {
			maxIndex = i
		}
	}

	root := &TreeNode{Val: nums[maxIndex]}
	root.Left = buildTree(nums[:maxIndex], start, maxIndex-1)
	root.Right = buildTree(nums[maxIndex+1:], maxIndex+1, end)

	return root
}
